default/drive_options.py

```diff
@@ -0,0 +1,50 @@
+import logging
+
+from django.http import HttpResponse, JsonResponse
+from google.auth.transport.requests import Request
+from google_auth_oauthlib.flow import InstalledAppFlow
+from googleapiclient.discovery import build
+from googleapiclient.http import MediaFileUpload
+from oauthlib.oauth2 import AccessDeniedError
+from ocdstoucan.settings import OCDS_TOUCAN_CREDENTIALS_DRIVE
+
+logger = logging.getLogger(__name__)
+
+SCOPES = 'https://www.googleapis.com/auth/drive.file'
+
+
+def upload_to_drive(filename, filepath, format=None, credentials=None):
+    try:
+        if not credentials or not credentials.valid:
+            if credentials and credentials.expired and credentials.refresh_token:
+                credentials.refresh(Request())
+            else:
+                flow = InstalledAppFlow.from_client_secrets_file(OCDS_TOUCAN_CREDENTIALS_DRIVE, SCOPES)
+                credentials = flow.run_local_server(port=0)
+        service = build('drive', 'v3', credentials=credentials)
+
+    except AccessDeniedError:
+        return HttpResponse("AccessDeniedError", status=400)
+
+    try:
+        if format == 'xlsx':
+            mimeType = 'application/vnd.google-apps.spreadsheet'
+        else:
+            mimeType = 'application/zip'
+
+        file_metadata = {
+            'name': filename,
+            'mimeType': mimeType
+        }
+        media = MediaFileUpload(filepath,
+                                mimetype=mimeType,
+                                resumable=True)
+        results = service.files().create(body=file_metadata, media_body=media, fields='id').execute()
+
+        return JsonResponse({
+            'name': filename,
+            'id': results["id"]
+        })
+
+    except (TypeError, Exception, IOError):
```

Comment:
Dos cosas aquí:

* Sería mejor dejar esto en `except:` solamente porque `Exception` es muy general. De hecho, tener estos tres tipos de excepciones ya es redundante porque los otros dos deberían heredar de `Exception`.
* Otra cosa es que podría ser mejor retornar un mensaje de error desde aquí en lugar de crear otro mensaje del lado cliente. Digo esto porque es más fácil crear y traducir los mensajes de este lado (usando `gettext()`) y mostrarlo en un solo espacio en la plantilla, que tener dos mensajes ocultos en la plantilla y tener que alternar entre mostrar uno y otro dependiendo de lo que viene del servidor. La segunda opción también es más flexible, por si en el futuro queramos agregar más mensajes de error no vamos a tener que alterar el cliente.

Reply:
Dejar solo en `except:` salta el error `E722 do not use bare 'except'`, lo deje en `except Exception:`. También retorno el mensaje traducido con `gettext()` que lo muestro así mismo en el cliente junto con un mensaje general del flow de autenticación del drive.

Reply:
Entiendo, dejar `Exception` está bien.